### PR TITLE
Update the select references UI

### PR DESCRIPTION
**Pull request by JR-G**
## Context

This PR is an update to the initial design of the _reference selection_ section in [#4793](https://github.com/DFE-Digital/apply-for-teacher-training/pull/4793).

### For a candidate who hasn't received feedback:

|Application review|Reference selection|
|---|---|
|image|Screenshot 2021-05-25 at 17 47 59|

### For a candidate who has feedback available for selection:

|Application review|Reference selection|Reference review|
|---|---|---|
|Screenshot 2021-05-25 at 17 48 23|Screenshot 2021-05-25 at 17 48 49|Screenshot 2021-05-25 at 17 52 48|

## Guidance to review

- Use the review app with the `reference_selection` feature flag on
- A candidate will need to have received two or more pieces of feedback in order for the selection view to render
- Some states still haven't been covered yet and will follow in a separate PR

This is a temporary fix to show the references in a bulleted list:
```
def reference_values
  list = '<ul class="govuk-list govuk-list--bullet">'.html_safe
  @references_selected.map do |reference|
    list <<
      '<li>'.html_safe <<
      "#{reference.referee_type.humanize} reference from #{reference.name}" <<
      '</li>'.html_safe
  end
  list + '</ul>'.html_safe
end
```

## Link to Trello card

https://trello.com/c/hZVGCcxd/3378-dev-choose-from-more-than-2-feedback-provided-references

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Comment by JR-G** [comment]
Currently, it doesn't retain your reference selection when you encounter the validation error:
image

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (paulrobertlloyd) [inline app/controllers/candidate_interface/references/selection_controller.rb:73]: This needs to generate an unordered list. I’m guessing this’ll require creating a component?

Current markup:

```html
<li>School based reference from Stanton Weimann</li>
<br>
<li>Character reference from Calvin Daugherty</li>
<br>
```

Desired markup:

```html
<ul class="govuk-list govuk-list--bullet">
  <li>School based reference from Stanton Weimann</li>
  <li>Character reference from Calvin Daugherty</li>
</ul>
```

File: app/views/candidate_interface/references/selection/new.html.erb
Comment (paulrobertlloyd) [inline app/views/candidate_interface/references/selection/new.html.erb:10]: 👀 

```suggestion
      <h1 class="govuk-heading-xl"><%= t('page_titles.references_select') %></h1>
```

File: app/views/candidate_interface/references/selection/new.html.erb
Comment (paulrobertlloyd) [inline app/views/candidate_interface/references/selection/new.html.erb:17]: Never quite sure what our preferred style is around indenting and multiple params…

```suggestion
            <%= f.govuk_check_box
              :selected,
              reference.id,
              label: { text: reference.name, size: 's' },
              hint: { text: "#{reference.referee_type.humanize} reference" }
            %>
```

File: app/views/candidate_interface/references/selection/new.html.erb
Comment (paulrobertlloyd) [inline app/views/candidate_interface/references/selection/new.html.erb:30]:
```suggestion
      <h1 class="govuk-heading-xl">Selected references</h1>
```

File: app/views/candidate_interface/references/selection/new.html.erb
Comment (paulrobertlloyd) [inline app/views/candidate_interface/references/selection/new.html.erb:36]:
```suggestion
      <%= govuk_link_to 'Return to application', candidate_interface_application_form_path, button: true %>
```

File: app/views/candidate_interface/references/selection/review.html.erb
Comment (paulrobertlloyd) [inline app/views/candidate_interface/references/selection/review.html.erb:13]:
```suggestion
<%= form_with(model: @section_complete_form, url: candidate_interface_complete_selected_references_path, method: :post) do |f| %>
  <h1 class="govuk-heading-xl"><%= t('page_titles.references') %></h1>

  <%= render(SummaryCardComponent.new(rows: @selected_references_rows)) %>

  <%= render CandidateInterface::CompleteSectionComponent.new(form: f) %>

  <%= f.govuk_submit t('save_and_continue') %>
<% end %>
```

**Review by paulrobertlloyd — CHANGES_REQUESTED** [review]
Few tweaks. Once this is merged in, we should review the content update where necessary.

File: app/views/candidate_interface/unsubmitted_application_form/show.html.erb
Comment (malcolmbaig) [inline app/views/candidate_interface/unsubmitted_application_form/show.html.erb:63]: Could make this an entry in this locale file https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/config/locales/en.yml#L233

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:30]: Needs to be the following to populate the radios:

`@section_complete_form = SectionCompleteForm.new(completed: current_application.references_completed)`

File: app/components/task_list_item_references_component.rb
Comment (malcolmbaig) [inline app/components/task_list_item_references_component.rb:7]: Given this component isn't used anywhere else (and so there's no need to customise the link text or path), I'd suggest keeping this simple for now and hardcoding them in the template rather than accepting parameters.

Comment (malcolmbaig) [reply app/components/task_list_item_references_component.rb:7]: (can always adopt the approach you've taken here if we need to display references in this way elsewhere in the sytem)

File: app/controllers/candidate_interface/references/review_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/review_controller.rb:139]: Can delete the `ApplicationReference.feedback_provided_but_not_selected` scope as well, as it's no longer used after this change.

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:9]: Can use `ApplicationForm#minimum_references_available_for_selection?` instead.

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:14]: As [above](https://github.com/DFE-Digital/apply-for-teacher-training/pull/4824/files#r640338586)

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:19]: Can use `@enough_references_provided` here

File: app/controllers/candidate_interface/references/review_controller.rb
Comment (malcolmbaig) [reply app/controllers/candidate_interface/references/review_controller.rb:139]: While we're at it, let's delete the `ApplicationReference.selected` I added earlier as well. Not needed, Rails already creates that scope!

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [reply app/controllers/candidate_interface/references/selection_controller.rb:73]: @JR-G let's amend `#reference_values` to something like the following and card up creating the new component.

```
        list = '<ul class="govuk-list govuk-list--bullet">'.html_safe
        @references_selected.each do |reference|
          list <<
            '<li>'.html_safe <<
            "#{reference.referee_type.humanize} reference from #{reference.name}" <<
            '</li>'.html_safe
        end
        list + '</ul>'.html_safe
```

File: spec/system/candidate_interface/references/reference_selection/candidate_selects_two_references_spec.rb
Comment (malcolmbaig) [inline spec/system/candidate_interface/references/reference_selection/candidate_selects_two_references_spec.rb:9]: Worth covering what happens when no references are received yet and the candidate navigates to the select references page.

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:44]: I think we need the additional check here https://github.com/DFE-Digital/apply-for-teacher-training/pull/4824/files#diff-9e94491041faa77499901ce7d101c0e3d79300775e64d0f6eb4d7a4831f29221L20 to validate selection of a radio button

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:47]: `and return`

will also need the same instance vars assigned as in the `review` action in order to render without error

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:58]: This else branch won't have any validation errors. It's really just here to handle stale form submissions, eg—they submit 'Yes I've completed this section' but they've deleted those references in another tab just before. Just redirecting to the selection review page is probably sufficient.

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (malcolmbaig) [inline app/controllers/candidate_interface/references/selection_controller.rb:40]: A useful addition would be a before_action filter that only applies to `review` and `complete`. This could check that enough references have been selected and if not, redirects the candidate to `candidate_interface_select_references_path`.

Comment (malcolmbaig) [reply app/controllers/candidate_interface/references/selection_controller.rb:40]: You can then remove the `else` branch at https://github.com/DFE-Digital/apply-for-teacher-training/pull/4824/files#diff-691fbbf565a6b43c0f77a4b5d263bba008f73d25169eba5b3b302942b2d33acfR67

File: app/controllers/candidate_interface/references/selection_controller.rb
Comment (stevehook) [inline app/controllers/candidate_interface/references/selection_controller.rb:3]: should we have a `render_404 unless FeatureFlag.active?...` in this controller?

**Review by stevehook — APPROVED** [review]
Have tested the review app and it works well as far as I can see. The code looks good and I think the feature flag is watertight (though I've suggested one minor change). So all good from a dev point of view as soon as Paul is happy.

**Review by paulrobertlloyd — APPROVED** [review]

**Pull request merged**
